design/design-doc-template-stacks-experience.md

```diff
@@ -0,0 +1,505 @@
+
+### Templating/configuration engine
+
+Template stacks will not be opinionated about which templating engine is
+used. We plan to support multiple configuration engines. The engine will
+be configurable by setting values in the `stack.yaml`.
+
+For some realistic examples of what a stack would look like when it uses
+different engines, see the [quick start example][quick-start-example].
+There are also some more details in the [stack yaml
+section](#the-stackyaml), and in the [helm charts
+section](#helm-charts).
+
+### Lifecycle hooks
+
+We expect to eventually support lifecycle hooks. See the [speculative
+design in the template stacks experience repo](https://github.com/suskin/template-stack-experience/tree/master/wordpress-workload/go-templates/lifecycle-hook)
+for more details about what that could look like.
+
+### Internal representation of templates
+
+See the [design doc on the internals of the Template Stack
+implementation](https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-template-stacks.md).
+
+### The template stack controller
+
+See the [design doc on the internals of the Template Stack
+implementation](https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-template-stacks.md).
+
+## Example use-cases
+
+See the [Template Stack
+Experience](https://github.com/suskin/template-stack-experience) code
+examples for some realistic examples of how different use-cases work
+with a realistic workload. For a coherent user scenario, see the [quick
+start example](https://github.com/suskin/template-stack-experience/tree/master/wordpress-workload/quick-start).
+
+Examples covered by the repository mostly use go templates as the
+templating engine. The following cases are covered by the examples:
+
+* No variables at all in a set of configurations.
+* Variables in configurations with default variable values.
+* Variables in configurations, with values specified by users when a
+  render is requested.
+* Resource packs.
+* Multiple CRDs defined in a single Stack.
+* Multiple rendered variants of a given configuration.
+* Composing configuration using multiple Stacks.
+* A user scenario of setting up cloud provider resources and deploying
+  an application bundled as a stack.
+
+### Helm charts
+
+To author a [template stack from a simple helm
+chart][helm-engine-example], these steps could be followed:
+
+1. Create and initialize a template stack project, with a helm flag:
+   `kubectl crossplane stack init --template`.
+2. Create a CRD: `kubectl crossplane stack crd init WordpressInstance
+   wordpressinstances wordpress.samples.stacks.crossplane.io`
+3. Put the chart's contents into the templates folder.
+4. Create a `stack.yaml` in the root of the repostory, and configure it
+   to use the templates for the created CRD.
+5. To set up default values:
+    1. Put the contents of the chart's `values.yaml` into the CRD's
```

Comment:
echoing @negz feedback from earlier, I wonder what tooling and experience improvements we can make around setting defaults in the CRDs, especially when coming from an existing helm chart, in addition to what @negz mentioned of tooling to define the CRD's OpenAPI schema.

I wonder if there should be any investment in Helm chart migration/import automation.  e.g. `kubectl crossplane stack init --template --from-helm-chart=some-chart-dir`?

Reply:
Yes, there should absolutely be an investment in improving the authoring and importing/converting stories. However, I believe we will develop a better idea of how we want to improve those as we build a working system and gain some experience with it. It seems out of scope for now.

If we are brainstorming, I imagine that a CRD, the start of a schema, some of the hook configuration, and defaults could all be generated from a helm chart.